crates/biome_js_semantic/src/format_semantic_model.rs

```diff
@@ -0,0 +1,201 @@
+                .print()
+                .expect("Expected a valid document")
+                .as_code(),
+        )
+    }
+}
+
+impl Format<FormatSemanticModelContext> for SemanticModel {
+    fn fmt(&self, f: &mut Formatter<FormatSemanticModelContext>) -> FormatResult<()> {
+        write!(f, [self.global_scope()])
+    }
+}
+
+enum ScopeOrBinding {
+    Scope(Scope),
+    Binding(Binding),
+}
+
+impl Format<FormatSemanticModelContext> for Scope {
+    fn fmt(&self, f: &mut Formatter<FormatSemanticModelContext>) -> FormatResult<()> {
+        let mut children = self
+            .descendents()
+            .skip(1) // skips the current scope
+            .map(ScopeOrBinding::Scope)
+            .chain(self.bindings().map(ScopeOrBinding::Binding))
+            .collect::<Vec<_>>();
+        // Yes it would be more efficient to just build the vector in the right order, but
+        // this is easier to read and maintain.
+        children.sort_by_key(|item| match item {
+            ScopeOrBinding::Scope(scope) => scope.range().start(),
+            ScopeOrBinding::Binding(binding) => binding.syntax().text_range_with_trivia().start(),
+        });
+
+        write!(
+            f,
+            [
+                text("Scope {"),
+                group(&block_indent(&format_args![
+                    format_with(|f| {
+                        let range = std::format!("{:?}", self.range());
+                        write!(
+                            f,
+                            [
+                                text("id: "),
+                                self.id,
+                                text(" @ "),
+                                dynamic_text(range.as_str(), TextSize::default()),
+                                hard_line_break()
+                            ]
+                        )?;
+
+                        Ok(())
+                    }),
+                    text("children ["),
+                    group(&block_indent(&format_with(|f| {
+                        if children.is_empty() {
+                            return write!(f, [text("No children")]);
+                        }
+                        for child in &children {
+                            match child {
+                                ScopeOrBinding::Scope(scope) => write!(f, [scope])?,
+                                ScopeOrBinding::Binding(binding) => write!(f, [binding])?,
+                            }
+                        }
+                        Ok(())
+                    }))),
+                    text("]"),
+                ])),
+                text("}"),
+                hard_line_break(),
+            ]
+        )?;
+        Ok(())
+    }
+}
+
+impl Format<FormatSemanticModelContext> for Binding {
+    fn fmt(&self, f: &mut Formatter<FormatSemanticModelContext>) -> FormatResult<()> {
+        write!(
+            f,
+            [
+                text("Binding {"),
+                group(&block_indent(&format_with(|f| {
```

Comment:
If you use this function, it means that you can store it inside a variable. That will improve readaibility